### PR TITLE
Kucoin: borrowMargin, repayMargin

### PR DESCRIPTION
**Pull request by Dan-krm**
Added borrowMargin and repayMargin to Kucoin:
### Cross borrowMargin:
```
node examples/js/cli kucoin borrowMargin USDT 100

kucoin.borrowMargin (USDT, 100)
2022-07-26T01:31:20.601Z iteration 0 passed in 917 ms

{
  id: '62df43e8cde9380001200377',
  currency: 'USDT',
  amount: 100,
  symbol: undefined,
  timestamp: 1658799080600,
  datetime: '2022-07-26T01:31:20.600Z',
  info: { orderId: '62df43e8cde9380001200377', currency: 'USDT' }
}
2022-07-26T01:31:20.601Z iteration 1 passed in 917 ms
```

### Isolated borrowMargin:
```
node examples/js/cli kucoin borrowMargin USDT 100 BTC/USDT '{"marginMode":"isolated"}'

kucoin.borrowMargin (USDT, 100, BTC/USDT, [object Object])
2022-07-26T01:39:39.294Z iteration 0 passed in 935 ms

{
  id: '62df45dab3c26b00012478e0',
  currency: 'USDT',
  amount: 100,
  symbol: 'BTC/USDT',
  timestamp: 1658799579293,
  datetime: '2022-07-26T01:39:39.293Z',
  info: {
    orderId: '62df45dab3c26b00012478e0',
    currency: 'USDT',
    actualSize: '100'
  }
}
2022-07-26T01:39:39.294Z iteration 1 passed in 935 ms
```

### Cross repayMargin:
```
node examples/js/cli kucoin repayMargin USDT 100.00008333

kucoin.repayMargin (USDT, 100.00008333)
2022-07-26T02:36:33.125Z iteration 0 passed in 502 ms

{
  id: undefined,
  currency: 'USDT',
  amount: 100.00008333,
  symbol: undefined,
  timestamp: 1658802993125,
  datetime: '2022-07-26T02:36:33.125Z',
  info: { code: '200000', data: null }
}
2022-07-26T02:36:33.125Z iteration 1 passed in 502 ms
```

### Isolated repayMargin:
```
node examples/js/cli kucoin repayMargin USDT 100.00008333 BTC/USDT '{"marginMode":"isolated"}'

kucoin.repayMargin (USDT, 100.00008333, BTC/USDT, [object Object])
2022-07-26T02:34:41.072Z iteration 0 passed in 513 ms

{
  id: undefined,
  currency: 'USDT',
  amount: 100.00008333,
  symbol: 'BTC/USDT',
  timestamp: 1658802881072,
  datetime: '2022-07-26T02:34:41.072Z',
  info: { code: '200000', data: null }
}
2022-07-26T02:34:41.072Z iteration 1 passed in 513 ms
```

### CODE REVIEW

File: js/kucoin.js [hunk 1]
```diff
@@ -2777,6 +2782,163 @@ module.exports = class kucoin extends Exchange {
...
+         * @see https://docs.kucoin.com/#post-borrow-order
+         * @see https://docs.kucoin.com/#isolated-margin-borrowing
+         * @param {string} code unified currency code of the currency to borrow
+         * @param {float} amount the amount to borrow
+         * @param {string|undefined} symbol unified market symbol, required for isolated margin
+         * @param {object} params extra parameters specific to the kucoin api endpoints
+         * @param {string} params.timeInForce either IOC or FOK
+         * @returns {object} a [margin loan structure]{@link https://docs.ccxt.com/en/latest/manual.html#margin-loan-structure}
+         */
+        await this.loadMarkets ();
```
Comment (carlosmiei) [inline js/kucoin.js:2799]: Can we add this exception?
```
{"code":"210014","msg":"Exceeds the max. borrowing amount, the remaining amount you can borrow: 0USDT"}
```

File: js/kucoin.js [hunk 2]
```diff
@@ -2777,6 +2782,163 @@ module.exports = class kucoin extends Exchange {
...
+        //     {
+        //         "code": "200000",
+        //         "data": {
+        //             "orderId": "62df44a1c65f300001bc32a8",
+        //             "currency": "USDT",
+        //             "actualSize": "100"
+        //         }
+        //     }
+        //
+        const data = this.safeValue (response, 'data', {});
```
Comment (carlosmiei) [inline js/kucoin.js:2844]: Even though our request failed, we're getting a "successful" response. We might need to revisit the error handling 
image

File: js/kucoin.js [hunk 3]
```diff
@@ -2777,6 +2782,163 @@ module.exports = class kucoin extends Exchange {
...
+        const transaction = this.parseMarginLoan (data, currency);
+        return this.extend (transaction, {
+            'amount': amount,
```
Comment (carlosmiei) [inline js/kucoin.js:2847]: For the isolated margin, shouldn't we be getting this from 'actualSize'? (Don't know if we can have some discrepancies between the requested amount and conceded amount)

File: js/kucoin.js [hunk 4]
```diff
@@ -2777,6 +2782,163 @@ module.exports = class kucoin extends Exchange {
...
+         * @see https://docs.kucoin.com/#quick-repayment
+         * @param {string} code unified currency code of the currency to repay
+         * @param {float} amount the amount to repay
+         * @param {string|undefined} symbol unified market symbol
+         * @param {object} params extra parameters specific to the kucoin api endpoints
+         * @param {string} params.sequence cross margin repay sequence, either 'RECENTLY_EXPIRE_FIRST' or 'HIGHEST_RATE_FIRST'
+         * @param {string} params.seqStrategy isolated margin repay sequence, either 'RECENTLY_EXPIRE_FIRST' or 'HIGHEST_RATE_FIRST'
+         * @returns {object} a [margin loan structure]{@link https://docs.ccxt.com/en/latest/manual.html#margin-loan-structure}
+         */
+        await this.loadMarkets ();
```
Comment (carlosmiei) [inline js/kucoin.js:2867]: Same issue as above
image

File: js/kucoin.js (hunk 3 above)
Comment (Dan-krm) [reply js/kucoin.js:2847]: I've set `'amount': this.safeNumber (info, 'actualSize'),` in parseMarginLoan will that override the extended amount?

File: js/kucoin.js [hunk 5]
```diff
@@ -2777,6 +2784,163 @@ module.exports = class kucoin extends Exchange {
...
+         * @see https://docs.kucoin.com/#isolated-margin-borrowing
+         * @param {string} code unified currency code of the currency to borrow
+         * @param {float} amount the amount to borrow
+         * @param {string|undefined} symbol unified market symbol, required for isolated margin
+         * @param {object} params extra parameters specific to the kucoin api endpoints
+         * @param {string} params.timeInForce either IOC or FOK
+         * @returns {object} a [margin loan structure]{@link https://docs.ccxt.com/en/latest/manual.html#margin-loan-structure}
+         */
+        await this.loadMarkets ();
+        const currency = this.currency (code);
```
Comment (carlosmiei) [inline js/kucoin.js:2802]: Can you please add this one as well?  (just discovered it testing cross)
`  {"code":"103000","msg":"Exceed the borrowing limit, the remaining borrowable amount is: 40USDT"}  `

File: js/kucoin.js (hunk 3 above)
Comment (carlosmiei) [reply js/kucoin.js:2849]: @Dan-krm no, it's the other way around, the second object you pass to `this.extend` ({'amount':amount, 'symbol': symbol} in this case) will override the first one, so you're overriding 'amount' and losing 'actualSize'.

Shouldn't we only override `symbol` considering that amount is already defined?
```Javascript
        return this.extend (transaction, {
            'symbol': symbol,
        })
```

Comment (Dan-krm) [reply js/kucoin.js]: I changed the extend statement to return 'symbol' for isolated and 'amount' for cross

Comment (carlosmiei) [reply js/kucoin.js]: Nice, thanks!

**Review by carlosmiei — APPROVED** [review]

**Review by carlosmiei — APPROVED** [review]

**Pull request merged**